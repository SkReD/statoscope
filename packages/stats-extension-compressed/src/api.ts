import { APIFactory } from '@statoscope/extensions';
import makeIndex, { IndexAPI } from '@statoscope/helpers/dist/indexer';
import { Format, Resource, Size } from './generator';

export type API = (compilationId: string, resourceId: string) => Size | null;

const makeAPI: APIFactory<Format, API> = (source) => {
  const sizeIndexes: Map<string, IndexAPI<string, Resource>> = new Map();

  for (const compilation of source.payload.compilations) {
    sizeIndexes.set(
      compilation.id,
      makeIndex((r) => r.id, compilation.resources),
    );
  }

  return (compilationId: string, resourceId: string): Size | null => {
    return sizeIndexes.get(compilationId)?.get(resourceId)?.size ?? null;
  };
};

export default makeAPI;
