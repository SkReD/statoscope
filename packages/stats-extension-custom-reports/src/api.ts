import { APIFactory } from '@statoscope/extensions';
import { RelationItem } from '@statoscope/types';
import makeIndex, { IndexAPI } from '@statoscope/helpers/dist/indexer';
import { Report } from '@statoscope/types/types/custom-report';
import { Format } from './generator';

export type API = {
  getReports: (
    compilationId?: string | null,
    type?: RelationItem['type'] | null,
    relatedId?: string | number,
  ) => Report<unknown, unknown>[];
  getById(id: string): Report<unknown, unknown> | null;
};

type CompilationIndexItem = {
  id: string | null;
  relationTypeIndex: IndexAPI<RelationItem['type'], TypeIndexItem>;
  itemIndex: IndexAPI<Report<unknown, unknown>['id'], Report<unknown, unknown>>;
};

type TypeIndexItem = {
  type: RelationItem['type'];
  relationIdIndex: IndexAPI<RelationItem['id'], IdIndexItem>;
  itemIndex: IndexAPI<Report<unknown, unknown>['id'], Report<unknown, unknown>>;
};

type IdIndexItem = {
  id: RelationItem['id'];
  index: IndexAPI<RelationItem['id'], Report<unknown, unknown>>;
};

const makeAPI: APIFactory<Format, API> = (source) => {
  const idIndex = makeIndex((entity: Report<unknown, unknown>) => entity.id);
  const compilationIndex = makeIndex((entity: CompilationIndexItem) => entity.id);

  for (const compilation of source.payload.compilations) {
    let compilationItem = compilationIndex.get(compilation.id);

    if (!compilationItem) {
      compilationItem = {
        id: compilation.id,
        relationTypeIndex: makeIndex((entity: TypeIndexItem) => entity.type),
        itemIndex: makeIndex((entity: Report<unknown, unknown>) => entity.id),
      };

      compilationIndex.add(compilationItem);
    }

    for (const item of compilation.reports) {
      compilationItem.itemIndex.add(item);
      idIndex.add(item);

      for (const related of item.relations ?? []) {
        let typeItem = compilationItem.relationTypeIndex.get(related.type);

        if (!typeItem) {
          typeItem = {
            type: related.type,
            relationIdIndex: makeIndex((entity: IdIndexItem) => entity.id),
            itemIndex: makeIndex((entity: Report<unknown, unknown>) => entity.id),
          };

          compilationItem.relationTypeIndex.add(typeItem);
        }

        let idItem = typeItem.relationIdIndex.get(related.id);

        if (!idItem) {
          idItem = {
            id: related.id,
            index: makeIndex((entity: Report<unknown, unknown>) => entity.id),
          };

          typeItem.relationIdIndex.add(idItem);
        }

        typeItem.itemIndex.add(item);
        idItem.index.add(item);
      }
    }
  }

  return {
    getReports: (
      compilationId: string | null = null,
      type?: RelationItem['type'] | null,
      relatedId?: string | number,
    ): Report<unknown, unknown>[] => {
      if (relatedId) {
        if (!type) {
          throw new Error('type must be specified');
        }

        return (
          compilationIndex
            .get(compilationId)
            ?.relationTypeIndex.get(type)
            ?.relationIdIndex.get(relatedId)
            ?.index.getAll() ?? []
        );
      }

      if (type) {
        return (
          compilationIndex
            .get(compilationId)
            ?.relationTypeIndex.get(type)
            ?.itemIndex.getAll() ?? []
        );
      }

      return compilationIndex.get(compilationId)?.itemIndex.getAll() ?? [];
    },
    getById(id: string): Report<unknown, unknown> | null {
      return idIndex.get(id);
    },
  };
};

export default makeAPI;
