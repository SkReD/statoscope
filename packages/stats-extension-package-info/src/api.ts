import { APIFactory } from '@statoscope/extensions';
import makeIndex, { IndexAPI } from '@statoscope/helpers/dist/indexer';
import { Format, Package, Instance } from './generator';

export type API = {
  getPackage: (compilationId: string | null, packageName: string) => Package | null;
  getInstance: (
    compilationId: string | null,
    packageName: string,
    instancePath: string,
  ) => Instance | null;
};

const makeAPI: APIFactory<Format, API> = (source) => {
  const packageIndexes: Map<string | null, IndexAPI<string, Package>> = new Map();
  const instanceIndexes: Map<Package, IndexAPI<string, Instance>> = new Map();

  for (const compilation of source.payload.compilations) {
    packageIndexes.set(
      compilation.id,
      makeIndex((r) => r.name, compilation.packages),
    );

    for (const packageItem of compilation.packages) {
      instanceIndexes.set(
        packageItem,
        makeIndex((r) => r.path, packageItem.instances),
      );
    }
  }

  return {
    getPackage(compilationId: string | null, packageId: string): Package | null {
      return (
        packageIndexes.get(compilationId)?.get(packageId) ??
        packageIndexes.get(null)?.get(packageId) ??
        null
      );
    },
    getInstance(
      compilationId: string | null,
      packageId: string,
      instancePath: string,
    ): Instance | null {
      const resolvedPackage =
        packageIndexes.get(compilationId)?.get(packageId) ??
        packageIndexes.get(null)?.get(packageId) ??
        null;

      if (!resolvedPackage) {
        return null;
      }

      return instanceIndexes.get(resolvedPackage)?.get(instancePath) ?? null;
    },
  };
};

export default makeAPI;
