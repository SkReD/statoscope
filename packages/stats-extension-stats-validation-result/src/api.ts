import { APIFactory } from '@statoscope/extensions';
import { RelationItem } from '@statoscope/types/types';
import { RuleDescriptor } from '@statoscope/types/types/validation/api';
import makeIndex, { IndexAPI } from '@statoscope/helpers/dist/indexer';
import { Format, Item } from './generator';

export type API = {
  getItems: (
    compilationId: string | null,
    type?: RelationItem['type'] | null,
    relatedId?: string | number,
  ) => Item[];
  getItemById(id: number): Item | null;
  getRule(id: string): RuleDescriptor | null;
};

type CompilationIndexItem = {
  id: string | null;
  relationTypeIndex: IndexAPI<RelationItem['type'], TypeIndexItem>;
  itemIndex: IndexAPI<Item['id'], Item>;
};

type TypeIndexItem = {
  type: RelationItem['type'];
  relationIdIndex: IndexAPI<RelationItem['id'], IdIndexItem>;
  itemIndex: IndexAPI<Item['id'], Item>;
};

type IdIndexItem = {
  id: RelationItem['id'];
  index: IndexAPI<RelationItem['id'], Item>;
};

const makeAPI: APIFactory<Format, API> = (source) => {
  const idIndex = makeIndex((entity: Item) => entity.id);
  const compilationIndex = makeIndex((entity: CompilationIndexItem) => entity.id);
  const ruleIndex = makeIndex((item) => item.name, source.payload.rules);

  for (const compilation of source.payload.compilations) {
    let compilationItem = compilationIndex.get(compilation.id);

    if (!compilationItem) {
      compilationItem = {
        id: compilation.id,
        relationTypeIndex: makeIndex((entity: TypeIndexItem) => entity.type),
        itemIndex: makeIndex((entity: Item) => entity.id),
      };

      compilationIndex.add(compilationItem);
    }

    for (const item of compilation.items) {
      compilationItem.itemIndex.add(item);
      idIndex.add(item);

      for (const related of item.related) {
        let typeItem = compilationItem.relationTypeIndex.get(related.type);

        if (!typeItem) {
          typeItem = {
            type: related.type,
            relationIdIndex: makeIndex((entity: IdIndexItem) => entity.id),
            itemIndex: makeIndex((entity: Item) => entity.id),
          };

          compilationItem.relationTypeIndex.add(typeItem);
        }

        let idItem = typeItem.relationIdIndex.get(related.id);

        if (!idItem) {
          idItem = {
            id: related.id,
            index: makeIndex((entity: Item) => entity.id),
          };

          typeItem.relationIdIndex.add(idItem);
        }

        typeItem.itemIndex.add(item);
        idItem.index.add(item);
      }
    }
  }

  return {
    getItems: (
      compilationId: string | null,
      type?: RelationItem['type'] | null,
      relatedId?: string | number,
    ): Item[] => {
      if (relatedId) {
        if (!type) {
          throw new Error('type must be specified');
        }

        return (
          compilationIndex
            .get(compilationId)
            ?.relationTypeIndex.get(type)
            ?.relationIdIndex.get(relatedId)
            ?.index.getAll() ?? []
        );
      }

      if (type) {
        return (
          compilationIndex
            .get(compilationId)
            ?.relationTypeIndex.get(type)
            ?.itemIndex.getAll() ?? []
        );
      }

      return compilationIndex.get(compilationId)?.itemIndex.getAll() ?? [];
    },
    getRule(id: string): RuleDescriptor | null {
      return ruleIndex.get(id)?.descriptor || null;
    },
    getItemById(id: number): Item | null {
      return idIndex.get(id);
    },
  };
};

export default makeAPI;
