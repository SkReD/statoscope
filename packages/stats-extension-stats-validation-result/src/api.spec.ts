import { TestEntry } from '@statoscope/types/types/validation/test-entry';
import { RuleDescriptor } from '@statoscope/types/types/validation/api';
import APIFactory from './api';
import Generator from './generator';

const generator = new Generator();

const fooRule: RuleDescriptor = {
  description: 'foo-rule description',
  package: {
    name: 'foo-rule',
    version: '1.0.0',
    homepage: 'https://statoscope.tech',
  },
};
const fooMessage: TestEntry = {
  message: 'foo-message',
  compilation: 'foo-compilation',
  related: [
    { type: 'module', id: 'foo-foo-module' },
    { type: 'entry', id: 'foo-foo-entry' },
  ],
};

generator.handleRule('foo-rule', fooRule);

generator.handleEntry('foo-rule', fooMessage);

generator.handleEntry(
  'foo-rule',
  {
    message: 'bar-message',
    compilation: 'foo-compilation',
    related: [
      { type: 'module', id: 'foo-foo-module' },
      { type: 'module', id: 'foo-bar-module' },
      { type: 'entry', id: 'foo-foo-entry' },
      { type: 'entry', id: 'foo-bar-entry' },
    ],
  },
  'warn',
);

generator.handleEntry('foo-rule', {
  message: 'bar-message',
  compilation: 'bar-compilation',
  related: [
    { type: 'module', id: 'bar-foo-module' },
    { type: 'module', id: 'bar-bar-module' },
    { type: 'entry', id: 'bar-foo-entry' },
    { type: 'entry', id: 'bar-bar-entry' },
  ],
});

const data = generator.get();
const api = APIFactory(data);

test('getItemById', () => {
  expect(api.getItemById(100)).toBeNull();
  expect(api.getItemById(0)).toMatchObject({
    id: 0,
    rule: 'foo-rule',
    message: 'foo-message',
  });
});

test('getRule', () => {
  expect(api.getRule('bar-rule')).toBeNull();
  expect(api.getRule('foo-rule')).toBe(fooRule);
});

describe('getItems', () => {
  test('should return nothing', () => {
    expect(api.getItems('foo', 'entry')).toStrictEqual([]);
    expect(api.getItems('foo', 'entry', 'baz')).toStrictEqual([]);
    expect(api.getItems('foo-compilation', 'entry', 'bar')).toStrictEqual([]);
  });

  test('should return all entry-items', () => {
    expect(api.getItems('foo-compilation', 'entry')).toMatchSnapshot();
    expect(api.getItems('bar-compilation', 'entry')).toMatchSnapshot();
  });

  test('should return entry-item with specific id', () => {
    expect(api.getItems('foo-compilation', 'entry', 'foo-foo-entry')).toMatchSnapshot();
    expect(api.getItems('bar-compilation', 'entry', 'bar-foo-entry')).toMatchSnapshot();
  });

  test('should return all module-items', () => {
    expect(api.getItems('foo-compilation', 'module')).toMatchSnapshot();
    expect(api.getItems('bar-compilation', 'module')).toMatchSnapshot();
  });

  test('should return all items', () => {
    expect(api.getItems('foo-compilation')).toMatchSnapshot();
    expect(api.getItems('bar-compilation')).toMatchSnapshot();
  });

  test('should return module-item with specific id', () => {
    expect(api.getItems('foo-compilation', 'module', 'foo-foo-module')).toMatchSnapshot();
    expect(api.getItems('foo-compilation', 'module', 'foo-bar-module')).toMatchSnapshot();
    expect(api.getItems('bar-compilation', 'module', 'bar-foo-module')).toMatchSnapshot();
  });
});
