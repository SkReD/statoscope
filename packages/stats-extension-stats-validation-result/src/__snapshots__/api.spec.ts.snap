// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getItems should return all entry-items 1`] = `
[
  {
    "details": [],
    "id": 0,
    "message": "foo-message",
    "related": [
      {
        "id": "foo-foo-module",
        "type": "module",
      },
      {
        "id": "foo-foo-entry",
        "type": "entry",
      },
    ],
    "rule": "foo-rule",
    "type": "error",
  },
  {
    "details": [],
    "id": 1,
    "message": "bar-message",
    "related": [
      {
        "id": "foo-foo-module",
        "type": "module",
      },
      {
        "id": "foo-bar-module",
        "type": "module",
      },
      {
        "id": "foo-foo-entry",
        "type": "entry",
      },
      {
        "id": "foo-bar-entry",
        "type": "entry",
      },
    ],
    "rule": "foo-rule",
    "type": "warn",
  },
]
`;

exports[`getItems should return all entry-items 2`] = `
[
  {
    "details": [],
    "id": 2,
    "message": "bar-message",
    "related": [
      {
        "id": "bar-foo-module",
        "type": "module",
      },
      {
        "id": "bar-bar-module",
        "type": "module",
      },
      {
        "id": "bar-foo-entry",
        "type": "entry",
      },
      {
        "id": "bar-bar-entry",
        "type": "entry",
      },
    ],
    "rule": "foo-rule",
    "type": "error",
  },
]
`;

exports[`getItems should return all items 1`] = `
[
  {
    "details": [],
    "id": 0,
    "message": "foo-message",
    "related": [
      {
        "id": "foo-foo-module",
        "type": "module",
      },
      {
        "id": "foo-foo-entry",
        "type": "entry",
      },
    ],
    "rule": "foo-rule",
    "type": "error",
  },
  {
    "details": [],
    "id": 1,
    "message": "bar-message",
    "related": [
      {
        "id": "foo-foo-module",
        "type": "module",
      },
      {
        "id": "foo-bar-module",
        "type": "module",
      },
      {
        "id": "foo-foo-entry",
        "type": "entry",
      },
      {
        "id": "foo-bar-entry",
        "type": "entry",
      },
    ],
    "rule": "foo-rule",
    "type": "warn",
  },
]
`;

exports[`getItems should return all items 2`] = `
[
  {
    "details": [],
    "id": 2,
    "message": "bar-message",
    "related": [
      {
        "id": "bar-foo-module",
        "type": "module",
      },
      {
        "id": "bar-bar-module",
        "type": "module",
      },
      {
        "id": "bar-foo-entry",
        "type": "entry",
      },
      {
        "id": "bar-bar-entry",
        "type": "entry",
      },
    ],
    "rule": "foo-rule",
    "type": "error",
  },
]
`;

exports[`getItems should return all module-items 1`] = `
[
  {
    "details": [],
    "id": 0,
    "message": "foo-message",
    "related": [
      {
        "id": "foo-foo-module",
        "type": "module",
      },
      {
        "id": "foo-foo-entry",
        "type": "entry",
      },
    ],
    "rule": "foo-rule",
    "type": "error",
  },
  {
    "details": [],
    "id": 1,
    "message": "bar-message",
    "related": [
      {
        "id": "foo-foo-module",
        "type": "module",
      },
      {
        "id": "foo-bar-module",
        "type": "module",
      },
      {
        "id": "foo-foo-entry",
        "type": "entry",
      },
      {
        "id": "foo-bar-entry",
        "type": "entry",
      },
    ],
    "rule": "foo-rule",
    "type": "warn",
  },
]
`;

exports[`getItems should return all module-items 2`] = `
[
  {
    "details": [],
    "id": 2,
    "message": "bar-message",
    "related": [
      {
        "id": "bar-foo-module",
        "type": "module",
      },
      {
        "id": "bar-bar-module",
        "type": "module",
      },
      {
        "id": "bar-foo-entry",
        "type": "entry",
      },
      {
        "id": "bar-bar-entry",
        "type": "entry",
      },
    ],
    "rule": "foo-rule",
    "type": "error",
  },
]
`;

exports[`getItems should return entry-item with specific id 1`] = `
[
  {
    "details": [],
    "id": 0,
    "message": "foo-message",
    "related": [
      {
        "id": "foo-foo-module",
        "type": "module",
      },
      {
        "id": "foo-foo-entry",
        "type": "entry",
      },
    ],
    "rule": "foo-rule",
    "type": "error",
  },
  {
    "details": [],
    "id": 1,
    "message": "bar-message",
    "related": [
      {
        "id": "foo-foo-module",
        "type": "module",
      },
      {
        "id": "foo-bar-module",
        "type": "module",
      },
      {
        "id": "foo-foo-entry",
        "type": "entry",
      },
      {
        "id": "foo-bar-entry",
        "type": "entry",
      },
    ],
    "rule": "foo-rule",
    "type": "warn",
  },
]
`;

exports[`getItems should return entry-item with specific id 2`] = `
[
  {
    "details": [],
    "id": 2,
    "message": "bar-message",
    "related": [
      {
        "id": "bar-foo-module",
        "type": "module",
      },
      {
        "id": "bar-bar-module",
        "type": "module",
      },
      {
        "id": "bar-foo-entry",
        "type": "entry",
      },
      {
        "id": "bar-bar-entry",
        "type": "entry",
      },
    ],
    "rule": "foo-rule",
    "type": "error",
  },
]
`;

exports[`getItems should return module-item with specific id 1`] = `
[
  {
    "details": [],
    "id": 0,
    "message": "foo-message",
    "related": [
      {
        "id": "foo-foo-module",
        "type": "module",
      },
      {
        "id": "foo-foo-entry",
        "type": "entry",
      },
    ],
    "rule": "foo-rule",
    "type": "error",
  },
  {
    "details": [],
    "id": 1,
    "message": "bar-message",
    "related": [
      {
        "id": "foo-foo-module",
        "type": "module",
      },
      {
        "id": "foo-bar-module",
        "type": "module",
      },
      {
        "id": "foo-foo-entry",
        "type": "entry",
      },
      {
        "id": "foo-bar-entry",
        "type": "entry",
      },
    ],
    "rule": "foo-rule",
    "type": "warn",
  },
]
`;

exports[`getItems should return module-item with specific id 2`] = `
[
  {
    "details": [],
    "id": 1,
    "message": "bar-message",
    "related": [
      {
        "id": "foo-foo-module",
        "type": "module",
      },
      {
        "id": "foo-bar-module",
        "type": "module",
      },
      {
        "id": "foo-foo-entry",
        "type": "entry",
      },
      {
        "id": "foo-bar-entry",
        "type": "entry",
      },
    ],
    "rule": "foo-rule",
    "type": "warn",
  },
]
`;

exports[`getItems should return module-item with specific id 3`] = `
[
  {
    "details": [],
    "id": 2,
    "message": "bar-message",
    "related": [
      {
        "id": "bar-foo-module",
        "type": "module",
      },
      {
        "id": "bar-bar-module",
        "type": "module",
      },
      {
        "id": "bar-foo-entry",
        "type": "entry",
      },
      {
        "id": "bar-bar-entry",
        "type": "entry",
      },
    ],
    "rule": "foo-rule",
    "type": "error",
  },
]
`;
